src/fiber/fiber.ml

```diff
@@ -420,10 +420,13 @@ module Mutex = struct
 
   let unlock t k =
     assert t.locked;
-    ( match Queue.pop t.waiters with
-    | None -> t.locked <- false
-    | Some next -> K.run next () );
-    k ()
+    match Queue.pop t.waiters with
+    | None ->
+      t.locked <- false;
+      k ()
+    | Some next ->
+      k ();
+      K.run next ()
```

Comment:
Not directly relevant to this commit, but I just realized, why don't we set `t.locked <- false` if the only element in the queue was popped?

The current code makes stuff like this possible:

```ocaml
let* () = Mutex.unlock t in
let* () = Mutex.unlock t  in
...
```

If there's more than one waiter.

I know that `unlock` isn't even available to the user, but this still seems a bit fishy perhaps?

Reply:
Because it's not unlocked. If there is only one element in the queue, we are passing the lock to them. There are similar problems even with previous implementation, that's I didn't expose `lock` and `unlock`.

For instance, one might do:

```ocaml
fork_and_join (fun () -> unlock t) (fun () -> unlock t)
```

The main issue here is that the only fiber that should be allowed to unlock a mutex is the one that locked it. Just like with normal threads. `with_lock` enforces that.